### PR TITLE
MC-33499: Grid Filter not working for Scheduled Export when selected any kind of Entity Type

### PR DESCRIPTION
**Pull request by engcom-Echo**
### Related Pull Requests
https://github.com/magento/partners-magento2ee/pull/446

### Fixed Issues (if relevant)

1. Fixes magento/partners-magento2ee#361

### Manual testing scenarios (*)

1. See: magento/partners-magento2ee#361

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Comment by engcom-Echo** [comment]
@magento run all tests

**Comment by engcom-Echo** [comment]
@magento run all tests

**Comment by engcom-Echo** [comment]
@magento run all tests

**Comment by engcom-Echo** [comment]
@magento run all tests

**Comment by engcom-Echo** [comment]
@magento run all tests

**Comment by engcom-Echo** [comment]
@magento run all tests

**Comment by engcom-Echo** [comment]
@magento run all tests

File: app/code/Magento/ImportExport/Model/ResourceModel/Export/AttributeGridCollection.php [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+ * See COPYING.txt for license details.
+ */
+declare(strict_types=1);
+
+namespace Magento\ImportExport\Model\ResourceModel\Export;
+
+use Magento\Framework\Data\Collection;
+
+/**
+ * Class AttributeGridCollection
```
Comment (zakdma) [inline app/code/Magento/ImportExport/Model/ResourceModel/Export/AttributeGridCollection.php]: Please provide meaningful description for class

File: app/code/Magento/ImportExport/Model/ResourceModel/Export/AttributeGridCollection.php [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+    /**
+     * @inheritDoc
+     *
+     * @param string $field
+     * @param array|int|string $condition
+     * @return $this
+     */
+    public function addFieldToFilter($field, $condition)
+    {
+        $value = (string)$condition['like'];
```
Comment (zakdma) [inline app/code/Magento/ImportExport/Model/ResourceModel/Export/AttributeGridCollection.php]: You should check that condition['like'] really exists. You should pass all not like conditions

File: app/code/Magento/ImportExport/Model/ResourceModel/Export/AttributeGridCollection.php [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+        $value = trim(trim($value, "'"), "%");
```
Comment (zakdma) [inline app/code/Magento/ImportExport/Model/ResourceModel/Export/AttributeGridCollection.php]: You can provide set of trimmed chars as second parameter of trim. No need to call it twice
https://www.php.net/manual/ru/function.trim.php

File: app/code/Magento/ImportExport/Model/ResourceModel/Export/AttributeGridCollection.php [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+     * @param  string $direction
+     * @return $this
+     */
+    public function setOrder($field, $direction = self::SORT_ORDER_DESC)
+    {
+        $this->_orderField = $field;
+        uasort($this->_items, [$this, 'compareAttributes']);
+
+        if ($direction == self::SORT_ORDER_DESC) {
+            $this->_items = array_reverse($this->_items, true);
```
Comment (zakdma) [inline app/code/Magento/ImportExport/Model/ResourceModel/Export/AttributeGridCollection.php]: This is redundant and make performance hit
Better create anonymous function for sorting and pass order to it using use statement.

File: app/code/Magento/ImportExport/Model/ResourceModel/Export/AttributeGridCollection.php [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+    /**
+     * Add select order
+     *
+     * @param  string $field
+     * @param  string $direction
+     * @return $this
+     */
+    public function setOrder($field, $direction = self::SORT_ORDER_DESC)
+    {
+        $this->_orderField = $field;
```
Comment (zakdma) [inline app/code/Magento/ImportExport/Model/ResourceModel/Export/AttributeGridCollection.php]: _orderField is not declared
Actually setOrder should not do sorting immediately but store sorting fields in $this->_orders
The same for addFieldToFilter
Actually filtering and sorting should perform in load (or loadData) method

File: app/code/Magento/ImportExport/Block/Adminhtml/Export/Filter.php [hunk 6]
```diff
@@ -426,7 +437,9 @@ public function getRowUrl($row)
      */
     public function prepareCollection(\Magento\Framework\Data\Collection $collection)
     {
-        $this->setCollection($collection);
-        return $this->getCollection();
+        $attributeGridCollection = $this->attributeGridCollectionFactory->create();
```
Comment (zakdma) [inline app/code/Magento/ImportExport/Block/Adminhtml/Export/Filter.php:440]: It is better to do this in controller and pass prepared collection to block.

**Review by zakdma — CHANGES_REQUESTED** [review]

**Comment by engcom-Echo** [comment]
@magento run all tests

File: app/code/Magento/ImportExport/Model/ResourceModel/Export/AttributeGridCollection.php [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+namespace Magento\ImportExport\Model\ResourceModel\Export;
+
+use Magento\Framework\Data\Collection;
+
+/**
+ * Association of attributes for grid
+ */
+class AttributeGridCollection extends Collection
+{
+
```
Comment (zakdma) [inline app/code/Magento/ImportExport/Model/ResourceModel/Export/AttributeGridCollection.php:19]: Remove redundant empty line

File: app/code/Magento/ImportExport/Block/Adminhtml/Export/Filter.php [hunk 8]
```diff
@@ -426,7 +437,9 @@ public function getRowUrl($row)
...
+        $gridCollection = $attributeGridCollection->setItems($collection->getItems());
+        $this->setCollection($gridCollection);
+        return $collection;
```
Comment (zakdma) [inline app/code/Magento/ImportExport/Block/Adminhtml/Export/Filter.php:444]: add empty line before return

File: app/code/Magento/ImportExport/Model/ResourceModel/Export/AttributeGridCollection.php [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+            $this->addItem($item);
+        }
+
+        return $this;
+    }
+
+    /**
+     * @inheritDoc
+     *
+     * @return int
```
Comment (zakdma) [inline app/code/Magento/ImportExport/Model/ResourceModel/Export/AttributeGridCollection.php]: Redundant @return annotation

File: app/code/Magento/ImportExport/Model/ResourceModel/Export/AttributeGridCollection.php [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+     */
+    public function getSize(): int
+    {
+        return count($this->getItems());
+    }
+
+    /**
+     * @inheritDoc
+     *
+     * @param string $field
```
Comment (zakdma) [inline app/code/Magento/ImportExport/Model/ResourceModel/Export/AttributeGridCollection.php]: Redundant @param and @return annotations

File: app/code/Magento/ImportExport/Model/ResourceModel/Export/AttributeGridCollection.php [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+            $this->addFilter($field, $value);
+        }
+
+        return $this;
+    }
+
+    /**
+     * @inheritDoc
+     *
+     * @param false $printQuery
```
Comment (zakdma) [inline app/code/Magento/ImportExport/Model/ResourceModel/Export/AttributeGridCollection.php]: Redundant @param and @return annotations

File: app/code/Magento/ImportExport/Model/ResourceModel/Export/AttributeGridCollection.php [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+    {
+        if (!$this->getFlag('isFilter') && !empty($this->_filters)) {
+            foreach ($this->_filters as $filter) {
+                foreach ($this->_items as $item) {
+                    $field = $item->getData($filter->getData('field')) ?? '';
+                    if (stripos($field, $filter->getData('value')) === false) {
+                        $this->removeItemByKey($item->getId());
+                    }
+                }
+                $this->setFlag('isFilter', true);
```
Comment (zakdma) [inline app/code/Magento/ImportExport/Model/ResourceModel/Export/AttributeGridCollection.php]: I suppose it should be moved after $this->_filters loop

File: app/code/Magento/ImportExport/Model/ResourceModel/Export/AttributeGridCollection.php [hunk 13]
```diff
@@ -0,0 +1,89 @@
...
+    }
+
+    /**
+     * @inheritDoc
+     *
+     * @param false $printQuery
+     * @param false $logQuery
+     * @return $this
+     */
+    public function load($printQuery = false, $logQuery = false)
```
Comment (zakdma) [inline app/code/Magento/ImportExport/Model/ResourceModel/Export/AttributeGridCollection.php]: Please extract filter code and sort code in separate private methods

File: app/code/Magento/ImportExport/Model/ResourceModel/Export/AttributeGridCollection.php [hunk 14]
```diff
@@ -0,0 +1,89 @@
...
+                    }
+                }
+                $this->setFlag('isFilter', true);
+            }
+        }
+
+        $sortOrder = $this->_orders['attribute_code'];
+        uasort($this->_items, function ($a, $b) use ($sortOrder) {
+            $cmp = strnatcmp($a->getData('attribute_code'), $b->getData('attribute_code'));
+            return $sortOrder === self::SORT_ORDER_ASC ? $cmp : -$cmp;
```
Comment (zakdma) [inline app/code/Magento/ImportExport/Model/ResourceModel/Export/AttributeGridCollection.php:103]: please add empty line before return

**Review by zakdma — CHANGES_REQUESTED** [review]

**Comment by engcom-Echo** [comment]
@magento run all tests

**Review by zakdma — APPROVED** [review]

**Comment by engcom-Echo** [comment]
@magento run all tests

**Comment by engcom-Echo** [comment]
@magento run all tests

**Comment by zakdma** [comment]
@magento import code to https://github.com/magento-tsg/magento2ce

**Comment by magento-engcom-team** [comment]
@zakdma the branch with code successfully imported into`magento-tsg/magento2ce` repository. Branch name: `imported-magento-magento2-31060`.

File: app/code/Magento/ImportExport/Model/ResourceModel/Export/AttributeGridCollection.php [hunk 15]
```diff
@@ -0,0 +1,109 @@
...
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function addFieldToFilter($field, $condition)
+    {
+        if (isset($condition['like'])) {
+            $value = trim((string)$condition['like'], "'%");
+            $value = str_replace('\\','',$value);
```
Comment (zakdma) [inline app/code/Magento/ImportExport/Model/ResourceModel/Export/AttributeGridCollection.php]: That is not enough/ Please refer to \Magento\Framework\DB\Helper\AbstractHelper::escapeLikeValue that used to escape values for like condition
You should create separate private method unescapeLikeCondition($condition) where do all the job
1) Trim ' and % as you did
2) Replace \_ to _
3) Replace \% to %

**Review by zakdma — CHANGES_REQUESTED** [review]

**Comment by engcom-Echo** [comment]
@magento run all tests

**Review by zakdma — APPROVED** [review]

**Comment by zakdma** [comment]
@magento import code to https://github.com/magento-tsg/magento2ce

**Comment by magento-engcom-team** [comment]
@zakdma the branch with code successfully imported into`magento-tsg/magento2ce` repository. Branch name: `imported-magento-magento2-31060`.

**Comment by engcom-Echo** [comment]
@magento run all tests

**Pull request merged**